### PR TITLE
Smarty - remove duplicate template dirs

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Followup to restore a missing uniqueness check from https://github.com/civicrm/civicrm-core/pull/30281/files#r1620547960

### CODE REVIEW

File: mixin/smarty-v2@1/mixin.php [hunk 1]
```diff
@@ -26,7 +26,7 @@
     $smarty = CRM_Core_Smarty::singleton();
     $v2 = isset($smarty->_version) && version_compare($smarty->_version, 3, '<');
     $templateDirs = (array) ($v2 ? $smarty->template_dir : $smarty->getTemplateDir());
-    $templateDirs = array_merge($newDirs, $templateDirs);
+    $templateDirs = array_unique(array_merge($newDirs, $templateDirs));
```
Comment (demeritcowboy) [inline mixin/smarty-v2@1/mixin.php:29]: Similar to the other PR, this won't uniqify on windows, e.g. I have two entries for the same template dir that look like

`...sites/default/files/civicrm\ext/thing/templates`
`...sites\default\files\civicrm\ext\thing\templates\`

And besides the slash direction, note the second has a trailing and the other doesn't.

Comment (demeritcowboy) [reply mixin/smarty-v2@1/mixin.php:29]: something like

```php
$templateDirs = array_merge($newDirs, $templateDirs);
$templateDirs = array_unique(array_map(function($v) {
  $v = str_replace(DIRECTORY_SEPARATOR, '/', $v);
  $v = rtrim($v, '/') . '/';
  return $v;
}, $templateDirs));
```

Comment (totten) [reply mixin/smarty-v2@1/mixin.php]: Seems reasonable to me. Haven't tested on Windows, but it still works on my Linux. Added that on top.

**Comment by totten** [comment]
Good catch @colemanw. MOP.

File: mixin/smarty-v2@1/mixin.php (hunk 1 above)
Comment (demeritcowboy) [reply mixin/smarty-v2@1/mixin.php]: Thanks. I suppose could micro-optimize it further if it still seems slow - e.g. it doesn't need to re-normalize the paths that are already in templateDirs that it normalized before.

**Comment by colemanw** [comment]
@demeritcowboy I'm curious to know what real-world scenario would cause the directory separator characters inside the `$templateDirs` array to be out-of-sync with the new items being added, given that items are always added in this way, wouldn't the separator character always be the same?

**Comment by eileenmcnaughton** [comment]
FWIW - I think the code merged for Smarty 5 upstream the other day (`prependTemplateDir()` )will be helpful in improving on this - but in future not now

**Comment by demeritcowboy** [comment]
@colemanw it looks like in my case the input is ALWAYS `...sites/default/files/civicrm\ext/thing/templates` but

smarty calls `$this->_realpath(rtrim($dir ?? '', "/\\") . DIRECTORY_SEPARATOR, true)` at some point to normalize everything

so the result of that, which is `...sites\default\files\civicrm\ext\thing\templates\`, is what our function sees the second time around when it calls $smarty->getTemplateDir() to get the current list, but

the newDirs contains the mixed slashes one ==> mismatch

**Pull request merged**